Custom dynamic uniforms

arose opened the pull request:
Following up on #6021 and #6150, here is another approach to have custom uniform values that are updated on a per-object basis during rendering.

Add some `.userData` to an object:

``` javascript
object.userData[ "pickingColor" ] = 1;
```

define a dynamic uniform to use it:

``` javascript
var pickingMaterial = new THREE.ShaderMaterial( {
    uniforms: {
        pickingColor: {
            type: "c",
            value: new THREE.Color(),
            valueFunc: function( value, camera, object ){
                value.setHex( object.userData.pickingColor );
                return value;
            }
        }
    },
    vertexShader: "...",
    fragmentShader: "..."
} );
pickingMaterial.hasDynamicUniforms = true;
```

WestLangley commented on examples/webgl_interactive_instances_gpu.html:
```diff
@@ -0,0 +1,438 @@
+				controls.panSpeed = 0.8;
+				controls.noZoom = false;
+				controls.noPan = false;
+				controls.staticMoving = true;
```
Do you mean `controls.staticMoving = false;`, since you have specified a damping factor? If you don't care about damping, then I expect you can remove your animation loop.

arose replied:
I simplified it a bit. The animation loop is in there mainly to sample rendering time.

WestLangley commented:
The 2000 materials share the same shader program, so the benefit here is on the CPU side? Correct?

arose commented:
> The 2000 materials share the same shader program, so the benefit here is on the CPU side? Correct?

Exactly. They share the same program hence it is not refreshed, but the material is (in `WebGLRenderer.setProgram`). Using a single material across many meshes (i.e. geometry instances) avoids refreshing the material in the render loop and speeds up initialization as only one material is created. When there are a lot of uniform in a material, especially matrix objects, it becomes really noticeable. For the included example, mesh creation time is halved on the machines I tested it.

Compared to the approach in #6150 I favor this one as it avoids to basically re-create the uniforms within the objects. For example, to have a uniform `modelViewMatrixInverseTranspose` or `modelViewProjectionMatrix` one can have a single matrix object per uniform and update it within `uniform.valueFunc( value, camera, object )` utilizing the given arguments.

My use case is in a molecular viewing app (http://arose.github.io/ngl/html/ngl.html?example=largeCapsid). There are 1680 instances of the shape in the center which can be individually picked (do a middle click to center on anything). Also, for the ray-casting shaders to do spheres and cylinders additional transformation matrices are used and just updated on a per-object basis in the render loop.

I haven't tried hardware instancing via `ANGLE_instanced_arrays` yet. As far as I understand it I could have attributes with that where a single value is repeated for a whole instance, which would be something like an instance-specific uniform, i suppose.

WestLangley commented:
Interesting demo! You do have 100m+ vertices, 6000+ drawcalls. It would be good if you could reduce the drawcalls count.

arose commented:
> Interesting demo! You do have 100m+ vertices, 6000+ drawcalls. It would be good if you could reduce the drawcalls count.

Thanks! Yes, certainly, there are 4 drawcalls per instance, 3 for the ribbons and 1 for the spheres. I could create the ribbons in one geometry to reduce the drawcalls. As for the 1680 instances I will look into the `ANGLE_instanced_arrays` extension (or I could create the coordinates explicitly and merge them into one big geometry?!).

Anyway. back to the PR. I would like to know if this is something that is useful enough to become a library feature? As to the implementation, `material.hasDynamicUniforms` should probably be calculated automatically.

gero3 commented:
I really like this idea.

But i would change valueFunc to ValueFunction and would change that it either defines value or valueFunction, not both at the same time, but I don't know the impact of this.

arose commented:
> But i would change valueFunc to ValueFunction and would change that it either defines value or valueFunction, not both at the same time, but I don't know the impact of this.

@gero I am not sure I understand what you mean exactly, could you please elaborate?

Currently uniforms are object literals:

```javascript
someUniform = {
    type: "v2",  // mapping THREE.Vector2 onto glsl vec2
    value: new THREE.Vector2()
}
```

I added `valueFunc` to the object literal. I did not reused the `value` property as that would required checking the type of it every time it is accessed, which seemed bad in term of performance and number of code changes (haven't tested the performance bit though). If there would be a dedicated uniform class, the interface could be better, I guess. Renaming `valueFunc` to `valueFunction` sounds good.

WestLangley commented:
@arose I wonder what the performance would be with one material and one geometry with vertex colors... Of course, you would not be able to share vertices in that case. Some experimentation with various approaches, including instancing, would be very interesting, IMO.

gero3 commented:
var pickingMaterial = new THREE.ShaderMaterial( {
        uniforms: {
            pickingColor: {
                type: "c",
                value: new THREE.Color(),
                valueFunc: function( value, camera, object ){
                    value.setHex( object.userData.pickingColor );
                    return value;
                }
            }
        },
        vertexShader: "...",
        fragmentShader: "..."
    } );
    pickingMaterial.hasDynamicUniforms = true;

I'd rather see following syntax becuase It forces people to reuse objects and also you were setting the property value twice in current form: Once in the function and once when the value is returned.

    var pickingMaterial = new THREE.ShaderMaterial( {
        uniforms: {
            pickingColor: {
                type: "c",
                value: new THREE.Color(),
                valueFunc: function( uniform, camera, object ){
                    uniform.value.setHex( object.userData.pickingColor );
                }
            }
        },
        vertexShader: "...",
        fragmentShader: "..."
    } );
    pickingMaterial.hasDynamicUniforms = true;

arose commented:
> I'd rather see following syntax becuase It forces people to reuse objects and also you were setting the property value twice in current form: Once in the function and once when the value is returned.

@gero I like it, the whole PR is about re-using and the clearer that is the better.

arose commented:
> Some experimentation with various approaches, including instancing, would be very interesting, IMO.

@WestLangley indeed, all on my list.

arose commented:
Expanded the demo and finally tried "real" instancing: http://arose.github.io/demo/dynamic-uniforms2/examples/webgl_interactive_instances_gpu.html.

* merged: single draw call, every attribute in a big buffer
* instanced: single draw call, `ANGLE_instanced_arrays`
* singleMaterial: multiple drawcalls, single material, color uniforms changed in the render loop
* multiMaterial: multiple drawcalls, multiple materials

Renamed `valueFunc` to `updateFunction`, first parameter is now the `uniform` and not the `value`.

When `ANGLE_instanced_arrays` is available (and you are prepared to write your own shader) it seems the best choice.

Thanks for your suggestions @WestLangley @gero3!

arose commented:
Oh, and there is a strange *bug* when you go from the "instanced" method back to the "merged" method, then all objects have the same color. Only the "merged" method is affected by that. I have no idea why this occurs. I see this on Chrome and Firefox (just one machine tested).

kintel commented:
+1
I just started hacking on something very similar when I discovered this.
I'll do some testing

mrdoob commented:
@kintel let me know what you find

kintel commented:
@arose Initial testing: Looks good!
One suggestion for a demo feature: If you use the Scene.overrideMaterial feature in the singlematerial case, you can get away with creating a single scene, i.e. no duplication of meshes, by e.g. adding an 'id' userdata to each object and bind that to the color uniform for the pickingMaterial.

kintel commented:
I've been testing some more and I like this a lot. I agree that hasDynamicUniforms could be auto-detected.

Suggestions for a convincing demo:
* Read a handful of geometry objects from file
* Add a transform animation to each mesh (just some arbitrary local coordinate system rotation)
* Create a bunch of meshes, assigning a color and ID to each one, reusing the geometries
* Implement picking on these dynamic meshes using the ID color technique (using overrideMaterial to do the picking)

arose commented:
Thank you, @kintel, for looking into this! The idea to employ overrideMaterial to half the mesh count sounds really great, I will gladly implement that for the demo.

I would do the automatic detection of dynamic uniforms within a material in WebGLRenderer#initMaterial and add it to material.properties, or is there any need in exposing it?

@kintel did you noticed any oddities in the demo when going from the "instanced" method back to the "merged" method? Like what I noted earlier?

kintel commented:
@arose I agree with the initMaterial and materialProperties suggestion. I don't see any immediate need for exposing it.

I see the same oddity on Chrome on Mac. Could you share any ideas you have to why this happens? I can take a look - just want to know what you've already tested..

kintel commented:
@arose Update: I see the bug on http://threejsworker.com/api/pullrequests/7048/examples/index.html#webgl_interactive_instances_gpu

..but when I build it myself it works fine.

arose committed:
dynamic uniforms

arose committed:
cleanup

arose committed:
automatic detection of dynamic uniforms

arose committed:
use overrideMaterial for picking (singleMaterial only)

arose commented:
@kintel yes, fixed for me too, very nice

So, I added automatic detection of dynamic uniforms and added the overrideMaterial approach to the demo, which (as expected) speeds up scene creation some more when comparing singleMaterial vs multiMaterial. For 5000 instances I get ~220 ms (singleMaterial, useOverrideMaterial) vs ~420 ms (singleMaterial) vs ~610 ms (multiMaterial).

kintel commented:
Thx - works for me.

arose committed:
more useful GUI, formatting cleanup

arose commented:
Added statistics to the demo GUI so that it is easier to compare the effects the different construction and rendering methods have.

@mrdoob any (final) thoughts on this?

threejsworker commented:
The examples of this pullrequest are now built and visible in threejsworker. To view them, go to the following link: 

 http://threejsworker.com/viewpullrequest.html#7048

threejsworker commented:
The examples of this pullrequest are now built and visible in threejsworker. To view them, go to the following link: 

 http://threejsworker.com/viewpullrequest.html#7048

threejsworker commented:
The examples of this pullrequest are now built and visible in threejsworker. To view them, go to the following link: 

 http://threejsworker.com/viewpullrequest.html#7048

threejsworker commented:
The examples of this pullrequest are now built and visible in threejsworker. To view them, go to the following link: 

 http://threejsworker.com/viewpullrequest.html#7048

threejsworker commented:
The examples of this pullrequest are now built and visible in threejsworker. To view them, go to the following link: 

 http://threejsworker.com/viewpullrequest.html#7048

threejsworker commented:
The examples of this pullrequest are now built and visible in threejsworker. To view them, go to the following link: 

 http://threejsworker.com/viewpullrequest.html#7048

threejsworker commented:
The examples of this pullrequest are now built and visible in threejsworker. To view them, go to the following link: 

 http://threejsworker.com/viewpullrequest.html#7048

threejsworker commented:
The examples of this pullrequest are now built and visible in threejsworker. To view them, go to the following link: 

 http://threejsworker.com/viewpullrequest.html#7048

threejsworker commented:
The examples of this pullrequest are now built and visible in threejsworker. To view them, go to the following link: 

 http://threejsworker.com/viewpullrequest.html#7048

threejsworker commented:
The examples of this pullrequest are now built and visible in threejsworker. To view them, go to the following link: 

 http://threejsworker.com/viewpullrequest.html#7048

threejsworker commented:
The examples of this pullrequest are now built and visible in threejsworker. To view them, go to the following link: 

 http://threejsworker.com/viewpullrequest.html#7048

threejsworker commented:
The examples of this pullrequest are now built and visible in threejsworker. To view them, go to the following link: 

 http://threejsworker.com/viewpullrequest.html#7048

threejsworker commented:
The examples of this pullrequest are now built and visible in threejsworker. To view them, go to the following link: 

 http://threejsworker.com/viewpullrequest.html#7048

arose commented:
@gero3 seems like threejsworker (which is awesome) is a bit overzealous here

gero3 commented:
I've cleaned it a bit. Sorry about it.

mrdoob commented:
Sorry for taking so long reviewing this @arose.

I like the simplicity. But, how about we implement it this way instead?

```javascript
function onUpload( object, camera ) {
    this.value.setHex( object.userData.pickingColor );
}

var pickingMaterial = new THREE.ShaderMaterial( {
    uniforms: {
        pickingColor: new THREE.Uniform( 'c', new THREE.Color() ).onUpload( onUpload ),
    },
    vertexShader: "...",
    fragmentShader: "..."
} );
```

It's actually the same thing but prettier API. If it looks ok to you I can merge the PR and prettify.

kintel commented:
@mrdoob fwiw, I'm happy with either approach.
I'm not sure about the method name _onUpload_ - a bit technical perhaps. ..or did you want to make it look similar to _texture.onUpdate_ ?

arose commented:
@mrdoob Now I am sorry for the delay. I am all for using a Uniform class. Feel free to merge and prettify. Please let me now if there is anything I could do.

mrdoob commented:
Sweet! Thanks!

mrdoob commented:
> I'm not sure about the method name onUpload - a bit technical perhaps.

Well, it's a `Uniform` that is being added to a `ShaderMaterial`, so I wouldn't think is that technical 😁

kintel commented:
@mrdoob The name just confused me a bit, as the function isn't called upon upload, but rather the shader is uploaded if the value returned by the changes (unless I misunderstood something)..

mrdoob commented:
I changed it to `onUpdate`, which is also a bit confusing, but can't think of a better name.

pailhead commented:
Hi, 

I can't find any of these examples, but i am super curious what this is all about.

kintel commented:
http://threejs.org/examples/webgl_interactive_instances_gpu.html

Somehow it looks like that didn't make it into the main example index..

pailhead commented:
found it, studying,


Does this basically save you from "pre-creating" N materials for N different colors, rather you give every object it's own value for the uniform, and it set's the uniform on the same material right before the drawcall?

kintel commented:
@pailhead Yes, that would be one use-case for this. It makes it possible to map to whatever uniform you want, instead of a few hardcoded ones.

pailhead commented:
Im a little bit confused with why you would want this specific functionality hardcoded in the library. Although i don't quite understand the difference between merging , and instancing.


I was wondering, in this context, what do you think about a more generalized entrypoint /callback for this kind of stuff. 

I am still trying to learn git, and i have no idea where my pull request or discussion went, but in general, in the webglrenderer, in renderObjects, before issuing a drawcall have a THREE.Object3D.onWillRender() fire. 

I'm using this to set up stencil states, not sure if anything changed with the newer versions but so far three exposed very few methods if any for accessing the stencil buffer. 




If i understand correctly though, this thing has to happen or prevent stuff from happening before the setProgram gets called?

pailhead commented:
i.e. if there is a generalized callback at a specific place in the render function, could this just be solved by 

`
myMesh.onWillRender = function(){
     this.material.uniforms.myUniform.value = this.customData; 
}.bind(myMesh);

`

while many meshes share the sam instance of the material?

arose commented:
Hi @pailhead, there was quite some discussion leading to this PR you might find interesting to read: #6010, #6021, #6150.

pailhead commented:
I've went through the other issues, but i'm still not sure if i understand why this has to be it's own specific thing.  Or maybe i'm confused with the terminology ie. `dynamic`.

Lets say we have a uniform, that cares about the screensize:

```
uSSize:{
  type:'v4',
  value:new THREE.Vector4(screen.w , screen.h , 1/screen.w , 1/screen.h )
}
```

I want it updated on window resize, so it sounds like "dynamic" is a good way to describe it. However, i don't need this happening every tick. Right now I might be lured to give it a callback, that maps the window size to the vec4 value, because it is dynamic, it does need to change throughout the apps lifecycle. 

The difference between this - a uniform that you want to update at some point, and this dynamic one is that the dynamic one gets updated before every drawcall within one frame, while the window size just sets the state once at some frame.

Here are some thoughts:

1. If these standalone meshes are so tightly coupled together with this specific type of uniform, and the fact that they all have to be aware that they are using the same instance of a shadermaterial, why not make a THREE.MeshCluster( geom , mat , instanceCount , individualDataArray ) or something along those lines. Or even adding all these meshes to one specific Scene , that has an override material, and is doing this type of managing. 

2. If more flexibility is required, why not a generalized callback, where you can do whatever you want with the objects, materials etc?

3. If this stays the same, "dynamic" should probably be given some more though. 



Of course, i may be completely missing the point, as i did find all of this super confusing :)

mrdoob commented:
@pailhead to be honest, it is a bit confusing for me too, and I do like your approach better, but at the moment I can't think of the repercussions so I'm curious to hear @arose thoughts.

I guess for your approach we should somehow notify the material/program that the uniform updated before rendering.

pailhead commented:
Im just thinking, ok so what if we push this optimization to the extreme, do we always have just one material, that sets itself as it encounters each object in the scene graph? I mean, this is almost like going backwards a step towards the state machine. While i'm up for any kind of optimization, this one doesn't seem like it's necessary, it's too specific to one use case, and is buried quite deep in the core of three. 

But, lets say that this stays, would another viable approach be to change the way the shader materials are constructed? I.e if you create 2000 materials with the same vertex and fragment shader, just cache one, and setup a different mechanism to have it called with the proper per object uniform.

arose commented:
> How much impact does bind() have when called on every object in every frame along with these checks, the array, the loop vs an already preconfigured and bound callback:

I don't know, but it seems not to be too bad. Anyway if I understand your code correctly you are proposing a mesh-based callback. As far as I can remember we had not considered that.

I agree with @mrdoob that there must be some mechanism to tell the program about a changed uniform. This might be the tricky part.

> I.e if you create 2000 materials with the same vertex and fragment shader, just cache one, and setup a different mechanism to have it called with the proper per object uniform.

Not sure if I understand. The materials programs are already cached on the renderer level. I tried to avoid creating the JavaScript material objects to have a faster startup

pailhead commented:
I'm trying to figure out what's confusing me :)


Does any of this, make any sense:

> If these standalone meshes are so tightly coupled together with this specific type of uniform, and the fact that they all have to be aware that they are using the same instance of a shadermaterial, why not make a THREE.MeshCluster( geom , mat , instanceCount , individualDataArray ) or something along those lines. Or even adding all these meshes to one specific Scene , that has an override material, and is doing this type of managing.

For example:

```javascript
    sequencedMeshes = new THREE.Scene()
    sequencedMeshes.overrideMaterial = mySharedMaterial;
    
    for ( var i = 0; i< instanceCount ; i++){
      var oneMesh = new THREE.Mesh( geom );
      var oneMesh.someCustomValue = Math.random();

      sequencedMeshes.add( oneMesh );
    }
```

There still needs to be a way to notify the override / shared material that it needs to take some data from the mesh, but the callback that im suggesting (or event) would take care of that. 

I guess what im saying is, yes, ok, this applies to uniforms in this particular use case, but the solution is very tightly coupled with the problem, and it seems to add a bit of complexity.

pailhead commented:
> Not sure if I understand. The materials programs are already cached on the renderer level. I tried to avoid creating the JavaScript material objects to have a faster startup

i'm not sure what happens when you call "new" in js, ie. if there was some factory class in three that creates materials with matFactory.createShaderMateraial(), you could avoid calling new, and just pass the same, single instance of a material to every mesh if it's the same shader. This is what you were trying to avoid correct? Calling `new THREE.ShaderMaterial({..})` thousands of times? 

You don't have to create a new instance of a material, i dont think that was ever a problem, you could just pass the same material when constructing a mesh, or none even, just use it as a scene override material.

The problem was how to update the uniforms. But since this is already stepping back from abstraction, towards webgl, i'd rather expose an entry point, where we can inject our own GL calls. Am i wrong in thinking that this is a similar problem (low level uniform management, vs actual GL calls)

tschw commented:
> How much impact does bind() have when called on every object in every frame along with these checks, the array, the loop vs an already preconfigured and bound callback?

You really don't want to use the `bind` method in performance-sensitive library code: It's *a whole lot* slower than using a function with an upvalue, where "a whole lot" refers to a browser-dependent ~100 - ~7000% time overhead per call.

See https://jsperf.com/bind-vs-emulate/41 and previous revisions.
